protobuf_serialization.nim

```diff
@@ -0,0 +1,376 @@
+    var value: int64
+  else:
+    var value: int32
+  var shiftAmount = 0
+
+  for _ in 0 ..< sizeof(T):
+    value += type(value)(bytes[outOffset]) shl shiftAmount
+    shiftAmount += 8
+    increaseBytesRead()
+
+  result = cast[T](value)
+
+proc get[T: SomeVarint](
+  bytes: var seq[byte],
+  ty: typedesc[T],
+  outOffset: var int,
+  outBytesProcessed: var int,
+  numBytesToRead = none(int)
+): T {.inline.} =
+  var bytesRead = 0
+  # Only up to 128 bits supported by the spec
+  when T is enum or T is char:
+    var value: type(ord(result))
+  elif T is bool:
+    var value: byte
+  else:
+    var value: T
+
+  var shiftAmount = 0
+  while true:
+    value += type(value)(bytes[outOffset] and 0b0111_1111) shl shiftAmount
+    shiftAmount += 7
+    if (bytes[outOffset] shr 7) == 0:
+      break
+    increaseBytesRead()
+
+  increaseBytesRead()
+
+  when ty is SomeSVarint:
+    if (value and type(value)(1)) != type(value)(0):
+      result = cast[T](not(value shr type(value)(1)))
+    else:
+      result = cast[T](value shr type(value)(1))
+  else:
+    result = T(value)
+
+proc checkType[T: SomeVarint](tyByte: byte, ty: typedesc[T], offset: int) {.inline.} =
+  let wireTy = wireType(tyByte)
+  if wireTy != Varint:
+    raise newException(UnexpectedTypeError, fmt"Not a varint at offset {offset}! Received a {wireTy}")
+
+proc checkType[T: SomeFixed](tyByte: byte, ty: typedesc[T], offset: int) {.inline.} =
+  let wireTy = wireType(tyByte)
+  if wireTy notin {Fixed32, Fixed64}:
+    raise newException(UnexpectedTypeError, fmt"Not a fixed32 or fixed64 at offset {offset}! Received a {wireTy}")
+
+proc checkType[T: SomeLengthDelimited](tyByte: byte, ty: typedesc[T], offset: int) {.inline.} =
+  let wireTy = wireType(tyByte)
+  if wireTy != LengthDelimited:
+    raise newException(UnexpectedTypeError, fmt"Not a length delimited value at offset {offset}! Received a {wireTy}")
+
+proc checkType[T: object](tyByte: byte, ty: typedesc[T], offset: int) {.inline.} =
+  let wireTy = wireType(tyByte)
+  if wireTy != LengthDelimited:
+    raise newException(UnexpectedTypeError, fmt"Not an object value at offset {offset}! Received a {wireTy}")
+
+proc get*[T: SomeLengthDelimited](
+  bytes: var seq[byte],
+  ty: typedesc[T],
+  outOffset: var int,
+  outBytesProcessed: var int,
+  numBytesToRead = none(int)
+): T {.inline.} =
+  var bytesRead = 0
+  let decodedSize = bytes.get(uint, outOffset, outBytesProcessed, numBytesToRead)
+  let length = decodedSize.int
+
+  when T is string:
+    result = newString(length)
+    for i in outOffset ..< (outOffset + length):
+      result[i - outOffset] = bytes[i].chr
+  elif T is cstring:
+    result = cast[cstring](bytes[outOffset ..< (outOffset + length)])
+  else:
+    result.setLen(length)
+    for i in outOffset ..< (outOffset + length):
+      result[i - outOffset] = type(result[0])(bytes[i])
+
+  increaseBytesRead(length)
+
+proc decodeField*[T: SomeFixed | SomeVarint | SomeLengthDelimited](
+  bytes: var seq[byte],
```

Comment:
I'd like to see decoders that are reading from a stream. The only tricky part will be decoding objects, because the fields there can appear in arbitrary order. To handle this, you'll use roughly the following loop:

```nim
let fields = T.fieldReadersTable(ProtoBufReader)
while not protoBufReader.stream.eof:
   let fieldId = protoBufReader.readFieldId
   let fieldReaderProc = fields.getFieldReader(fieldId)
   fieldReaderProc(decodedObject, protoBufReader)
```

You'll find a very similar example in the [JsonReader implementation](https://github.com/status-im/nim-json-serialization/blob/master/json_serialization/reader.nim#L261-L286). In general, I would suggest studying the JSON library a bit more and following its structure with separate reader and writer modules.